client/src/bundles/UserForms.js

```diff
@@ -1,738 +1,742 @@
+    this.$el.on('userform.form.valid', (e, fieldId) => {
+      self.errorContainer.removeStepLink(fieldId);
+    });
+
+    this.$el.validate(this.validationOptions);
+
+    // Ensure checkbox groups are validated correctly
+    $('.optionset.requiredField input').each(() => {
+      $(this).rules('add', {
+        required: true,
+      });
+    });
+
+    return this;
+  }
+
+  /*
+   * Default options for step validation. These get extended in main().
+   */
+  UserForm.prototype.validationOptions = {
+    ignore: ':hidden,ul',
+    errorClass: 'error',
+    errorElement: 'span',
+    errorPlacement: (error, element) => {
+      error.addClass('message');
+
+      if (element.is(':radio') || element.parents('.checkboxset').length > 0) {
+        error.insertAfter(element.closest('ul'));
+      } else if (element.parents('.checkbox').length > 0) {
+        error.insertAfter(element.next('label'));
+      } else {
+        error.insertAfter(element);
+      }
+    },
+    invalidHandler: (event, validator) => {
+      // setTimeout 0 so it runs after errorPlacement
+      setTimeout(() => {
+        validator.currentElements.filter('.error').first().focus();
+      }, 0);
+    },
+    // Callback for handling the actual submit when the form is valid.
+    // Submission in the jQuery.validate sence is handled at step level.
+    // So when the final step is submitted we have to also check all previous steps are valid.
+    submitHandler: (form) => {
+      let isValid = true;
+
+      // Validate the current step
+      if (userform.currentStep) {
+        userform.currentStep.valid = $(form).valid();
+      }
+
+      // Check for invalid previous steps.
+      $.each(userform.steps, (i, step) => {
+        if (!step.valid && !step.conditionallyHidden()) {
+          isValid = false;
+          userform.errorContainer.addStepLink(step);
+        }
+      });
+
+      if (isValid) {
+        // When using the "are you sure?" plugin, ensure the form immediately submits.
+        $(form).removeClass('dirty');
+
+        form.submit();
+      } else {
+        userform.errorContainer.show();
+      }
+    },
+    // When a field becomes valid.
+    success: (error) => {
+      const errorId = $(error).attr('id');
+      const fieldId = errorId.substr(0, errorId.indexOf('-error')).replace(/[\\[\\]]/, '');
+
+      // Remove square brackets since jQuery.validate.js uses idOrName,
+      // which breaks further on when using a selector that end with
+      // square brackets.
+
+      error.remove();
+
+      // Pass the field's ID with the event.
+      userform.$el.trigger('userform.form.valid', [fieldId]);
+    },
+  };
+
+  /**
+   * @func UserForm.addStep
+   * @param {object} step - An instance of FormStep.
+   * @desc Adds a step to the UserForm.
+   */
+  UserForm.prototype.addStep = (step) => {
+    // Make sure we're dealing with a form step.
+    if (!step instanceof FormStep) {
+      return;
+    }
+
+    step.id = this.steps.length;
```

Comment:
I'd suggest putting a `// eslint-disable-next-line no-param-reassign` above this line for now, it's quite involved to be trying to not reassign param props this early in the migration